### PR TITLE
Fix reflect typeMembers to return all members

### LINKED ISSUE #14902: Can't get inherited types from Quotes
## Compiler version

2ec9c52

## Minimized code

Having a symbol of a class, we can retrieve from compiler API the information about:
- declared methods
- declared + inherited methods
- declared fields
- declared + inherited fields
- declared types

However, we can't retrieve information about **inherited types**.

## Expectation

Looking at the API, I'm not sure if it's not a oversight since there's a naming convention:
 - `declaredMethods` - methods declared in a class
 - `methodMembers` - methods declared in a class and inherited
 - `declaredFields` - fields declared in a class
 - `fieldMembers` - fields declared in a class and inherited

And the method: `typeMembers` breaks the convention since it returns only declared types.

Aren't we missing this method?

### PR DESCRIPTION
**Pull request by nicolasstucki**
Fixes #14902

### CODE REVIEW

**Comment by pikinier20** [comment]
I've pushed a commit repairing Scaladoc tests

**Comment by nicolasstucki** [comment]
>  I've pushed a commit repairing Scaladoc tests

Did not notice this commit and force pushed the fix I had. @pikinier20 is this the same fix that you did?

**Comment by pikinier20** [comment]
> > I've pushed a commit repairing Scaladoc tests
> 
> Did not notice this commit and force pushed the fix I had. @pikinier20 is this the same fix that you did?

Yes, it is. Sorry for messing up :)

**Comment by nicolasstucki** [comment]
Note that the other fixes in the documentation are in #15034

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -2619,13 +2619,19 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
...
           }.toList
 
-        def memberType(name: String): Symbol = typeMember(name)
-        def typeMember(name: String): Symbol =
+        def memberType(name: String): Symbol =
           self.unforcedDecls.find(sym => sym.name == name.toTypeName)
+        def typeMember(name: String): Symbol =
+          lookupPrefix.allMembers.iterator.map(_.symbol).collect {
+            case sym if sym.isType && sym.name.toString == name => sym.asType
+          }.toList.headOption.getOrElse(dotc.core.Symbols.NoSymbol)
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2627]: Traversing all members to find one with a specific name is really inefficient. Could you instead do:
```suggestion
          lookupPrefix.memberBasedOnFlags(name.toTypeName)
```

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -2619,13 +2619,19 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
...
 
-        def memberTypes: List[Symbol] = typeMembers
-        def typeMembers: List[Symbol] =
+        def memberTypes: List[Symbol] =
           self.unforcedDecls.filter(_.isType)
+        def typeMembers: List[Symbol] =
+          lookupPrefix.allMembers.iterator.map(_.symbol).collect {
+            case sym if sym.isType => sym.asType
+          }.toList
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2634]: In general, we shouldn't use `allMembers` for performance reasons, either use `memberDenots` directly or other methods built on top of `memberDenots`:
```suggestion
          lookupPrefix.typeMembers
```

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: updated

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: updated

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 3]
```diff
@@ -2619,13 +2619,15 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
             case sym if sym.isType => sym.asType
           }.toList
 
-        def memberType(name: String): Symbol = typeMember(name)
-        def typeMember(name: String): Symbol =
+        def memberType(name: String): Symbol =
           self.unforcedDecls.find(sym => sym.name == name.toTypeName)
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2623]: I guess this was present before, but any reason we use unforcedDecls here? That could also miss declarations.

Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2623]: I do not belive there was a particular reason. Which method would you use instead?

Comment (smarter) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2623]: `decls`

Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: done

**Review by smarter — APPROVED** [review]

**Pull request merged**